main.go

```diff
@@ -33,54 +39,69 @@ func main() {
 	}
 	log.Event(nil, "config on startup", log.Data{"config": cfg})
 
-	router := mux.NewRouter()
-
-	dc := dataset.NewAPIClient(cfg.DatasetAPIURL)
+	ctx := context.Background()
 
-	buildTime, err := strconv.ParseInt(BuildTime, 10, 64)
-	if err != nil {
-		log.Event(nil, "failed to parse build time", log.Error(err), log.Data{"buildTime": BuildTime})
-	}
-	versionInfo := healthcheck.CreateVersionInfo(
-		time.Unix(buildTime, 0),
+	versionInfo, err := healthcheck.NewVersionInfo(
+		BuildTime,
 		GitCommit,
 		Version,
 	)
+	if err != nil {
+		log.Event(ctx, "failed to create service version information", log.Error(err))
+		os.Exit(1)
+	}
 
-	hc := healthcheck.Create(versionInfo, cfg.HealthCheckCritialTimeout, cfg.HealthCheckInterval, nil)
+	dc := dataset.NewAPIClient(cfg.DatasetAPIURL)
 
+	hc := healthcheck.New(versionInfo, cfg.HealthCheckCritialTimeout, cfg.HealthCheckInterval)
+	if err = hc.AddCheck("dataset API", dc.Checker); err != nil {
+		log.Event(ctx, "failed to add dataset API checker", log.Error(err))
+		os.Exit(1)
+	}
+
+	router := mux.NewRouter()
 	routes.Init(router, cfg, hc, dc)
 
 	s := server.New(cfg.BindAddr, router)
 
-	ctx := context.Context(context.Background())
-	hc.Start(ctx)
-
 	go func() {
 		if err := s.ListenAndServe(); err != nil {
 			log.Event(nil, "error starting http server", log.Error(err))
-			os.Exit(1)
+			return
 		}
 	}()
 
-	for {
-		select {
-		case <-signals:
-			log.Event(nil, "os signal received")
-			gracefulShutdown(cfg, s, hc)
-		}
+	hc.Start(ctx)
+
+	// Block until a fatal error occurs
+	select {
+	case signal := <-signals:
+		log.Event(ctx, "quitting after os signal received", log.Data{"signal": signal})
 	}
-}
 
-func gracefulShutdown(cfg *config.Config, s *server.Server, hc healthcheck.HealthCheck) {
-	log.Event(nil, fmt.Sprintf("shutdown with timeout: %s", cfg.GracefulShutdownTimeout))
+	log.Event(ctx, fmt.Sprintf("shutdown with timeout: %s", cfg.GracefulShutdownTimeout))
+
+	// give the app `Timeout` seconds to close gracefully before killing it.
 	ctx, cancel := context.WithTimeout(context.Background(), cfg.GracefulShutdownTimeout)
-	if err := s.Shutdown(ctx); err != nil {
-		log.Event(nil, "failed to gracefully shutdown http server", log.Error(err))
+
+	go func() {
+		log.Event(ctx, "stop health checkers")
+		hc.Stop()
+
+		if err := s.Shutdown(ctx); err != nil {
+			log.Event(ctx, "failed to gracefully shutdown http server", log.Error(err))
+		}
+
+		cancel() // stop timer
+	}()
+
+	// wait for timeout or success (via cancel)
+	<-ctx.Done()
+	if ctx.Err() == context.DeadlineExceeded {
+		log.Event(ctx, "context deadline exceeded", log.Error(ctx.Err()))
+	} else {
+		log.Event(ctx, "graceful shutdown complete", log.Data{"context": ctx.Err()})
 	}
-	log.Event(nil, "graceful shutdown of http server complete", nil)
-	hc.Stop()
-	log.Event(nil, "shutdown complete", nil)
-	cancel()
+
```

Comment:
👇 Should the os exit code always be 1?

Reply:
You are right, it should be 0 as it successfully shutdown